gluonnlp/data/sampler.py

```diff
@@ -82,6 +79,141 @@ def _bucket_average_lengths(bucket_sample_ids, seq_lengths):
     return bucket_average_lengths
 
 
+class BucketScheme(object):
+    r"""Base class for generating bucket keys.
+    """
+    def __call__(self, max_lengths, min_lengths, num_buckets):
+        """Generate bucket keys based on the lengths of sequences and number of buckets.
+
+        Parameters
+        ----------
+        max_lengths : int of list of int
+            Maximum of lengths of sequences.
+        min_lengths : int of list of int
+            Minimum of lengths of sequences.
+        num_buckets : int
+            Number of buckets
+        """
+        raise NotImplementedError
+
+
+class ConstWidthBucket(BucketScheme):
+    r""" Buckets with constant width.
+    """
+    def __call__(self, max_lengths, min_lengths, num_buckets):
+        r"""This generate bucket keys given that all the buckets have the same width.
+
+        Parameters
+        ----------
+        max_lengths : int of list of int
+            Maximum of lengths of sequences.
+        min_lengths : int of list of int
+            Minimum of lengths of sequences.
+        num_buckets : int
+            Number of buckets
+        """
+        if not isinstance(max_lengths, INT_TYPES):
+            bucket_width_l = [max((max_len - min_len) // num_buckets, 1)
+                              for max_len, min_len in
+                              zip(max_lengths, min_lengths)]
+            bucket_keys = \
+                [tuple(max(max_len - i * width, min_len) for max_len, min_len, width in
+                       zip(max_lengths, min_lengths, bucket_width_l))
+                 for i in range(num_buckets)]
+        else:
+            bucket_width = max((max_lengths - min_lengths) // num_buckets, 1)
+            bucket_keys = [max(max_lengths - i * bucket_width, min_lengths)
+                           for i in range(num_buckets)]
+        return bucket_keys
+
+
+class LinearWidthBucket(BucketScheme):
+    r""" Buckets with linearly increasing width.
+    """
+    def __call__(self, max_lengths, min_lengths, num_buckets):
+        r"""This implements a scheme for linearly increasing bucket width:
+                :math:`w_i = \alpha * i + 1` for all :math:`i \geq 1`.
```

Comment:
put this in class doc. method doc should explain what the method's arguments and return values are